Add method to add instructions to a DAGCircuit from an iterator of PackedInstruction

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->
Solves #13003 

### Summary
Tracked by #13001 and preceeded by #12975
The following commits aim to add a `DAGCircuit::extend()` to add a chain of `DAGOpNodes` to the DAGCircuit based on a sequence of `PackedInstruction`.

### Details and comments
These commits add the following two methods:
- `DAGCircuit::extend()` to add a sequence of `PackedInstruction` continously avoiding having to re-add each bit/var's output node each time an instruction is added. The addition of the output node is only performed once all instructions have been added correctly to the DAGCircuit.

### Known issues:
- At the moment, I'm still thinking of ways of handling vars correctly as we need to keep track of each bit/var's output node and vars are currently represented as `PyObject` which are not hashable.
    - ~~One possible solution would be to use a `PyDict` however the structures that are being stored within this mapping are rust-native (`NodeIndex`, `Wire`)~~.
    - ~~Another solution would be to somehow internalize these structures, which I'm still thinking on how to do.~~
- Feel free to comment with any ideas.

### Blockers
- [x] #12975 (will rebase once merged)
- [x] #12550

raynelfss committed:
Initial: Add add_from_iter method to DAGCircuit
- Introduce a method that adds a chain of `PackedInstruction` continuously avoiding the re-linking of each bit's output-node until the very end of the iterator.
   - TODO: Add handling of vars
- Add header for a `from_iter` function that will create a `DAGCircuit` based on a chain of `PackedInstruction`.

raynelfss committed:
Fix: leverage new methods in layers
- Fix incorrect re-insertion of last_node.

raynelfss committed:
Fix: Keep track of Vars for add_from_iter
- Remove `from_iter`

raynelfss committed:
Fix: Incorrect modification of last nodes in `add_from_iter`.
- Use `entry` api to either modify or insert a value if missing.

raynelfss committed:
Fix: Cycling edges in when adding vars.
- A bug that adds duplicate edges to vars has been temporarily fixed. However, the root of this problem hasn't been found yet. A proper fix is pending. For now skip those instances.

raynelfss committed:
Fix: Remove set collecting all nodes to be connected.
- A set collecting all the new nodes to connect with a new node was preventing additional wires to connect to subsequent nodes.

raynelfss committed:
Fix: Adapt to #13033

ElePT commented:
I believe that the PR doesn't have any open blocker. I will remove the on hold label. Let me know if I missed something.

raynelfss committed:
Refactor: `add_from_iter` is now called `extend` to stick with `Rust` nomenclature.

raynelfss commented:
> I believe that the PR doesn't have any open blocker. I will remove the on hold label. Let me know if I missed something.

I was about to do that :) Thank you

ElePT commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -6347,6 +6345,139 @@ impl DAGCircuit {
+                vars_last_nodes.set_item(var, new_node.index())?;
+                if var_last_node == new_node {
+                    // TODO: Fix instances of duplicate nodes for Vars
+                    continue;
```
If I am not mistaken it looks like you were able to store the var keys as PyDicts, right? Is there something else to fix or was the TODO addressed?

raynelfss replied:
You are correct, I must have forgotten to remove it.

ElePT commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -6347,6 +6345,139 @@ impl DAGCircuit {
     }
+
+    /// Adds valid instances of [PackedInstruction] to the back of the Circuit.
+    pub fn extend<I>(&mut self, py: Python, iter: I) -> PyResult<Vec<NodeIndex>>
```
```suggestion
    /// Extends the DAG with valid instances of [PackedInstruction]
    pub fn extend<I>(&mut self, py: Python, iter: I) -> PyResult<Vec<NodeIndex>>
```

ElePT reviewed (commented):
Thanks @raynelfss! This PR is quite useful for transpiler passes that build new instructions (like unitary synthesis). I have been using these changes in my PR and everything has worked as expected so far, so you can have that as a "unit test" data point. I might be overseeing something but I would approve this PR after addressing a few minor (and quite superficial) comments.

raynelfss committed:
Fix docstring

- Caught by @ElePT

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

raynelfss committed:
Fix: Remove duplicate vars check

raynelfss committed:
Merge branch 'main' into rusty-dag-add-from-iter

ElePT reviewed (dismissed):
LGTM

ElePT reviewed (commented):
Actually, I think I have found an issue when trying to add instructions that have no clbits. When I initialize a packed instruction with `clbits: target_dag.cargs_interner.get_default()`,  I then get a `the caller is responsible for only using interner keys from the correct interner` error. I will try to come up with a minimal example that reproduces the problem.

ElePT commented:
The issue was fixed by #13092, it was unrelated to the changes in this PR :)

raynelfss committed:
Merge branch 'main' into rusty-dag-add-from-iter

kevinhartman commented:
I can give this a quick look as well if it's helpful (I see it is not in the merge queue yet)

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -6347,6 +6345,135 @@ impl DAGCircuit {
+        let vars_last_nodes: Bound<PyDict> = PyDict::new_bound(py);
+
+        // Store new nodes to return
+        let mut new_nodes = vec![];
```
As an optimization, you can use `iter.size_hint()` here to reserve an appropriately sized vector if whatever was passed for `iter` happens to provide `Some(upper_bound)` in the tuple returned by `size_hint`.

raynelfss replied:
In this case, since we don't exactly know the type of iterator the received collection will turn into, we can't obtain a size hint until we consume the collection into the iterator. This should be fine as long as we use the iterator right after.

kevinhartman commented on crates/circuit/src/dag_circuit.rs:
```diff
@@ -6347,6 +6345,135 @@ impl DAGCircuit {
+                let qubit_last_node = if let Some(node) = qubit_last_nodes.remove(qubit) {
+                    node
+                } else {
+                    let output_node = self.qubit_io_map[qubit.0 as usize][1];
```
Can you add a comment here to explain that what this code is doing is removing the output node edge the first time a qubit is used in `iter`, and that we re-add them after processing all instructions?

It also may be cleaner to use the `entry` API up on line 6393 to get a mutable reference to the value instead of removing `qubit` from `qubit_last_nodes` only to add it back with a different value, but I think a comment would at least be helpful.

kevinhartman reviewed (commented):
This is looking good, thanks for adding this @raynelfss!

The one larger point I think we should start to be weary of is introducing new methods that aren't covered by any of our existing testing. We've operated in a bit of a free-for-all the last few months while porting things to Rust, under the assumption that our existing Python test code would provide sufficient coverage. 

IMO, this new method is sufficiently complex and should be unit tested. I'm not sure if we've decided on an approach to exercise the Rust-only API of our `pyclass`s. Perhaps @mtreinish has opinions / suggestions?

mtreinish commented:
> IMO, this new method is sufficiently complex and should be unit tested. I'm not sure if we've decided on an approach to exercise the Rust-only API of our pyclasss. Perhaps @mtreinish has opinions / suggestions?

In general having rust unit tests is possible and we have them already in place where it is possible/makes sense for PRs like this with new rust specific functionality. See: https://github.com/Qiskit/qiskit/commit/86c63eb3d719a129dbcb92c56e9a67954030b803 for a recent example (which is already setup to run in CI: https://dev.azure.com/qiskit-ci/qiskit/_build/results?buildId=60452&view=logs&j=b9878877-bc64-52fc-fc03-577dd0aa7cb4&t=2cc4852e-f347-585e-188c-b93a0bf20b0f&l=363 ). But what we will struggle with here is that this new method takes a `py` token and we don't have a mechanism to interact with python via rust unit tests.

I would agree we should have this exercised somehow in an ideal world. But I think we probably are ok to let this slide if we can't come up with a solution for unit testing with a python dependency. Given that there are several PRs actively based on top of this and those are exercising the code and tested via Python AFAICT.

raynelfss committed:
Fix: Corrections from code review.
- Use Entry API to modify last nodes in the var.
- Build new_nodes with an allocated vec.
- Add comment explaining the removal of the edge between the output node and its predecessor.

raynelfss committed:
Fix: Improper use of `Entry API`.
- Use `or_insert_with` instead of `or_insert` to perform actions before inserting a value.

kevinhartman reviewed (approved):
LGTM, thanks for the changes!
